Use different exit code if updates are disabled

joschi opened the pull request:
# Summary

If asdf-vm was installed with a package manager and the user doesn't have the necessary permissions to update it with `asdf update`, asdf-vm emits an informational message and exits with exit code 1.

This makes it hard to programmatically detect whether the update failed or wasn't even attempted because it's not possible.

With this change, asdf-vm would exit with the exit code 2 if updates are disabled instead of exit code 1, which signals an error during update.

Refs r-darwish/topgrade#367

joschi commented:
The reason for the failing build on AppVeyor seems unrelated to this PR.

Stratus3D commented:
Thanks for the PR @joschi ! I like these changes. And yes, the AppVeyor builts are going to be for Windows support, which isn't yet working.

I see you updated the tests, but I think it might make more sense to have separate tests for each possible exit code (1 and 2). Would it be difficult to add a new unit test for the exit code of 1?

joschi commented:
@Stratus3D Thanks for your quick response!

I will think of a way to test for the different scenarios when failing to run `asdf update`.

In the linked issue at `topgrade`, we also discussed introducing a kind of dry-run flag for `asdf update`:
https://github.com/r-darwish/topgrade/issues/367#issuecomment-598192568

Does that sound useful for you or would you prefer signaling the inability to perform an update with the exit code of the `asdf update` command exclusively?

Stratus3D commented:
@joschi good question. I guess my first thought would be to keep things as simple as possible. We want users to be able to get the info they need, but we also want our API (the `asdf` commands) to be as simple as possible. I like exit codes because of their simplicity. They are easy to implement and are usually enough to give the user all the information they need.

I'm not sure there is a clear benefit to a dry-run flag. Updating is as simple as pulling down a newer tag with Git or using some operating system-specific package manager to update asdf.

Is there a clear use case for a dry-run flag that could not be handled by the `asdf update` as you have it here (with two exit codes)? . And if so can you explain what that scenario would be?

joschi commented:
@Stratus3D I think you're right and we should keep it as simple as possible.

I've changed the exit code for disabled updates to 42 in order not to conflict with other conventions (http://www.tldp.org/LDP/abs/html/exitcodes.html) and updated the tests.

Stratus3D commented:
Thanks @joschi ! The code looks good. Thanks for taking the time to write the unit tests.
